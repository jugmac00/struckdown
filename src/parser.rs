use std::iter;
use std::ops::Range;

use itertools::Either;
use lazy_static::lazy_static;
use pulldown_cmark as cm;
use regex::Regex;

use crate::event::{Alignment, Attrs, Event, EventType, Location, Str, Tag};

lazy_static! {
    static ref TEXT_ROLE_RE: Regex = Regex::new(r"\{([^\r\n\}]+)\}$").unwrap();
    static ref CODE_ROLE_RE: Regex = Regex::new(r"^\{([^\r\n\}]+)\}(?:\s+(.*?))$").unwrap();
    static ref HEADING_ID_RE: Regex = Regex::new(r"\s+\{#([^\r\n\}]+)\}\s*$").unwrap();
}

/// Reads until the end of an image tag all embedded content as raw string.
///
/// We do this because in markdown/cmark the alt text of an image is in fact
/// supporting a lot of markdown syntax but it usually gets rendered out into
/// an alt attribute.  Because of that we normalize this into text during parsing
/// already so that stream processors don't need to deal with this oddity.
fn read_raw_alt<'a, 'b, I: Iterator<Item = (cm::Event<'b>, Range<usize>)>>(
    iter: &'a mut I,
) -> String {
    let mut depth = 1;
    let mut rv = String::new();
    while let Some((event, _)) = iter.next() {
        match event {
            cm::Event::Start(..) => depth += 1,
            cm::Event::End(..) => depth -= 1,
            cm::Event::Text(text) => rv.push_str(&text),
            cm::Event::Code(code) => rv.push_str(&code),
            cm::Event::SoftBreak | cm::Event::HardBreak => rv.push('\n'),
            _ => {}
        }
        if depth == 0 {
            break;
        }
    }
    rv
}

/// A trailer is information that gets attached to the start tag when the end
/// tag is emitted.
///
/// Trailers are supported internally on all tags for which [`tag_supports_trailers`]
/// returns `true`.
pub enum Trailer<'data> {
    /// Defines the id attribute via trailer.
    Id(Str<'data>),
}

/// Checks if a tag supports trailers.
///
/// Currently all headlines are the only tags supporting trailers.
fn tag_supports_trailers(tag: Tag) -> bool {
    match tag {
        Tag::Heading1 => true,
        Tag::Heading2 => true,
        Tag::Heading3 => true,
        Tag::Heading4 => true,
        Tag::Heading5 => true,
        Tag::Heading6 => true,
        _ => false,
    }
}

/// Parses a string into an event stream with trailers.
///
/// This is normally not used as the `parse` function will already buffer
/// as necessary to automatically attach the trailers to the start tags.
fn parse_with_trailers<'data>(
    s: &'data str,
) -> impl Iterator<Item = (Event, Option<Trailer<'data>>)> {
    let mut opts = cm::Options::empty();
    opts.insert(cm::Options::ENABLE_TABLES);
    opts.insert(cm::Options::ENABLE_STRIKETHROUGH);
    opts.insert(cm::Options::ENABLE_TASKLISTS);
    opts.insert(cm::Options::ENABLE_FOOTNOTES);

    let parser = cm::Parser::new_with_broken_link_callback(s, opts, None);
    let mut iter = parser.into_offset_iter().peekable();
    let mut tag_stack = vec![];
    let mut pending_role = None;
    let mut pending_trailer = None;

    iter::from_fn(move || {
        let mut trailer = None;

        if let Some((event, range)) = iter.next() {
            // inefficient way to find the location
            let mut loc = Location {
                offset: range.start,
                len: range.end - range.start,
                line: s[..range.start].chars().filter(|&c| c == '\n').count() + 1,
                column: match s[..range.start].rfind('\n') {
                    Some(nl) => range.start - nl - 1,
                    None => range.start,
                },
            };

            // simple events
            let ty = match event {
                cm::Event::Start(cm_tag) => {
                    let mut attrs = Attrs::default();
                    let tag = match cm_tag {
                        cm::Tag::Paragraph => Tag::Paragraph,
                        cm::Tag::Heading(1) => Tag::Heading1,
                        cm::Tag::Heading(2) => Tag::Heading2,
                        cm::Tag::Heading(3) => Tag::Heading3,
                        cm::Tag::Heading(4) => Tag::Heading4,
                        cm::Tag::Heading(5) => Tag::Heading5,
                        cm::Tag::Heading(6) => Tag::Heading6,
                        cm::Tag::Heading(_) => unreachable!(),
                        cm::Tag::BlockQuote => Tag::BlockQuote,
                        cm::Tag::CodeBlock(kind) => match kind {
                            cm::CodeBlockKind::Fenced(lang) => {
                                let lang: Str<'_> = lang.into();
                                if let Some(m) = CODE_ROLE_RE.captures(lang.as_str()) {
                                    let g1 = m.get(1).unwrap();
                                    let g2 = m.get(2).unwrap();
                                    attrs.role = Some(lang.slice(g1.start(), g1.end()));
                                    attrs.argument = Some(lang.slice(g2.start(), g2.end()));
                                    Tag::Directive
                                } else {
                                    attrs.argument = Some(lang);
                                    Tag::FencedCode
                                }
                            }
                            cm::CodeBlockKind::Indented => Tag::IndentedCode,
                        },
                        cm::Tag::List(None) => Tag::UnorderedList,
                        cm::Tag::List(Some(start)) => {
                            attrs.start = Some(start as u32);
                            Tag::OrderedList
                        }
                        cm::Tag::Item => Tag::ListItem,
                        cm::Tag::FootnoteDefinition(id) => {
                            attrs.id = Some(id.into());
                            Tag::FootnoteDefinition
                        }
                        cm::Tag::Table(alignments) => {
                            attrs.alignments = Some(
                                alignments
                                    .into_iter()
                                    .map(|cm_align| match cm_align {
                                        cm::Alignment::None => Alignment::None,
                                        cm::Alignment::Left => Alignment::Left,
                                        cm::Alignment::Center => Alignment::Center,
                                        cm::Alignment::Right => Alignment::Right,
                                    })
                                    .collect(),
                            );
                            Tag::Table
                        }
                        cm::Tag::TableHead => Tag::TableHead,
                        cm::Tag::TableRow => Tag::TableRow,
                        cm::Tag::TableCell => Tag::TableCell,
                        cm::Tag::Emphasis => Tag::Emphasis,
                        cm::Tag::Strong => Tag::Strong,
                        cm::Tag::Strikethrough => Tag::Strikethrough,
                        cm::Tag::Link(_, target, title) => {
                            attrs.target = Some(target.into());
                            if !title.is_empty() {
                                attrs.title = Some(title.into());
                            }
                            Tag::Link
                        }
                        cm::Tag::Image(_, target, title) => {
                            // images are special in that we downgrade them from
                            // tags to toplevel events to not have to deal with
                            // nested text.
                            let alt = read_raw_alt(&mut iter);
                            return Some((
                                Event::new(
                                    EventType::Image {
                                        target: target.into(),
                                        alt: if alt.is_empty() {
                                            None
                                        } else {
                                            Some(alt.into())
                                        },
                                        title: if title.is_empty() {
                                            None
                                        } else {
                                            Some(title.into())
                                        },
                                    },
                                    loc,
                                ),
                                None,
                            ));
                        }
                    };
                    tag_stack.push(tag);
                    EventType::StartTag { tag, attrs }
                }
                cm::Event::End(_) => {
                    trailer = pending_trailer.take();
                    EventType::EndTag {
                        tag: tag_stack.pop().unwrap(),
                    }
                }
                cm::Event::Text(text) => {
                    let mut text: Str<'_> = text.into();

                    // handle roles
                    if let Some(&(cm::Event::Code(_), _)) = iter.peek() {
                        if let Some(m) = TEXT_ROLE_RE.captures(text.as_str()) {
                            let g0 = m.get(0).unwrap();
                            let g1 = m.get(1).unwrap();

                            // adjust the span of the text to not include the role.
                            let column_adjustment = g0.end() - g0.start();
                            loc.len -= column_adjustment;
                            pending_role =
                                Some((text.slice(g1.start(), g1.end()), column_adjustment));
                            text = text.slice(0, g0.start());
                        }
                    }

                    // handle explicitly defined IDs for headlines
                    if let Some(&(cm::Event::End(cm::Tag::Heading(_)), _)) = iter.peek() {
                        if let Some(m) = HEADING_ID_RE.captures(text.as_str()) {
                            let g0 = m.get(0).unwrap();
                            let g1 = m.get(1).unwrap();

                            // adjust the span of the text to not include the role.
                            let column_adjustment = g0.end() - g0.start();
                            loc.len -= column_adjustment;
                            pending_trailer = Some(Trailer::Id(text.slice(g1.start(), g1.end())));
                            text = text.slice(0, g0.start());
                        }
                    }

                    EventType::Text { text }
                }
                cm::Event::Code(value) => {
                    // if there is a pending role then we're not working with a
                    // code block, but an interpreted text one.
                    if let Some((role, column_adjustment)) = pending_role.take() {
                        loc.column -= column_adjustment;
                        loc.offset -= column_adjustment;
                        loc.len += column_adjustment;
                        EventType::InterpretedText {
                            text: value.into(),
                            role: role.into(),
                        }
                    } else {
                        EventType::InlineCode { code: value.into() }
                    }
                }
                cm::Event::Html(html) => EventType::RawHtml { html: html.into() },
                cm::Event::FootnoteReference(target) => EventType::FootnoteReference {
                    target: target.into(),
                },
                cm::Event::SoftBreak => EventType::SoftBreak,
                cm::Event::HardBreak => EventType::HardBreak,
                cm::Event::Rule => EventType::Rule,
                cm::Event::TaskListMarker(checked) => EventType::Checkbox { checked },
            };

            Some((Event::new(ty, loc), trailer))
        } else {
            None
        }
    })
}

/// Recursively attaches trailers to start tags.
fn buffer_for_trailers<'data, I>(event: Event<'data>, iter: &mut I) -> Vec<Event<'data>>
where
    I: Iterator<Item = (Event<'data>, Option<Trailer<'data>>)>,
{
    let mut buffer = vec![event];
    let mut depth = 1;

    while let Some((event, trailer)) = iter.next() {
        // keep track of the tag depth
        match event.ty() {
            &EventType::StartTag { tag, .. } => {
                if tag_supports_trailers(tag) {
                    buffer.extend(buffer_for_trailers(event, iter));
                    continue;
                } else {
                    depth += 1;
                }
            }
            &EventType::EndTag { .. } => depth -= 1,
            _ => {}
        }
        buffer.push(event);

        // attach an end tag trailer to the start tag if needed.
        if depth == 0 {
            if let EventType::StartTag { attrs, .. } = buffer[0].ty_mut() {
                match trailer {
                    Some(Trailer::Id(new_id)) => {
                        attrs.id = Some(new_id);
                    }
                    None => {}
                }
            }
            break;
        }
    }

    buffer
}

/// Parses structured cmark into an event stream.
pub fn parse(s: &str) -> impl Iterator<Item = Event> {
    let mut iter = parse_with_trailers(s);
    iter::from_fn(move || {
        if let Some((event, _)) = iter.next() {
            if let &EventType::StartTag { tag, .. } = event.ty() {
                if tag_supports_trailers(tag) {
                    return Some(Either::Left(
                        buffer_for_trailers(event, &mut iter).into_iter(),
                    ));
                }
            }
            Some(Either::Right(iter::once(event)))
        } else {
            None
        }
    })
    .flatten()
}
