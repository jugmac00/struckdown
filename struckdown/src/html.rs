//! Implements an HTML renderer.
use std::borrow::Cow;
use std::collections::HashMap;
use std::io::{self, Write};

use serde::{Deserialize, Serialize};
use v_htmlescape::escape;

use crate::event::{
    Alignment, AnnotatedEvent, Attrs, CheckboxEvent, CodeBlockEvent, DirectiveEvent, EndTagEvent,
    ErrorEvent, Event, FootnoteReferenceEvent, ImageEvent, InlineCodeEvent, InterpretedTextEvent,
    RawHtmlEvent, StartTagEvent, Str, Tag, TextEvent,
};

/// Customizes the HTML rendering.
#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(default)]
pub struct HtmlRendererOptions {
    /// When enabled `_foo_` renders into underlines.
    pub render_underlines: bool,
    /// The class to emit for footnote references.
    pub footnote_reference_class: String,
    /// The class to emit for footnote definitions.
    pub footnote_definition_class: String,
    /// The initial level for headlines
    pub initial_headline_level: usize,
}

impl Default for HtmlRendererOptions {
    fn default() -> Self {
        Self::new()
    }
}

impl HtmlRendererOptions {
    /// Creates the default configuraiton for the renderer options.
    pub fn new() -> HtmlRendererOptions {
        HtmlRendererOptions {
            render_underlines: false,
            footnote_reference_class: "footnote-reference".into(),
            footnote_definition_class: "footnote-definition".into(),
            initial_headline_level: 1,
        }
    }
}

/// Object capable of rendering events to HTML.
pub struct HtmlRenderer<'data, 'options, F> {
    out: F,
    footnotes: HashMap<Str<'data>, usize>,
    options: &'options HtmlRendererOptions,
}

impl<'data, 'options, F: Write> HtmlRenderer<'data, 'options, F> {
    /// Creates a new renderer that writes into a writer.
    pub fn new(out: F, options: &'options HtmlRendererOptions) -> HtmlRenderer<'data, 'options, F> {
        HtmlRenderer {
            out,
            footnotes: HashMap::new(),
            options,
        }
    }

    /// Consumes the writer and returns the inner file.
    pub fn into_writer(self) -> F {
        self.out
    }

    fn is_block_tag(&self, tag: Tag) -> bool {
        match tag {
            Tag::Paragraph => true,
            Tag::Heading1 => true,
            Tag::Heading2 => true,
            Tag::Heading3 => true,
            Tag::Heading4 => true,
            Tag::Heading5 => true,
            Tag::Heading6 => true,
            Tag::BlockQuote => true,
            Tag::OrderedList => true,
            Tag::UnorderedList => true,
            Tag::ListItem => true,
            Tag::FootnoteDefinition => true,
            Tag::Table => true,
            Tag::TableHead => true,
            Tag::TableRow => true,
            Tag::TableCell => true,
            Tag::Emphasis => false,
            Tag::EmphasisAlt => false,
            Tag::Strong => false,
            Tag::Strikethrough => false,
            Tag::Link => false,
            Tag::TableHeader => true,
            Tag::TableBody => true,
            Tag::Container => true,
            Tag::Span => false,
        }
    }

    fn tag_to_html_tag(&self, tag: Tag) -> &'static str {
        match tag {
            Tag::Paragraph => "p",
            Tag::Heading1
            | Tag::Heading2
            | Tag::Heading3
            | Tag::Heading4
            | Tag::Heading5
            | Tag::Heading6 => {
                let level = tag.header_level().unwrap();
                let target_level = (self.options.initial_headline_level.saturating_sub(1)) + level;
                match target_level {
                    1 => "h1",
                    2 => "h2",
                    3 => "h3",
                    4 => "h4",
                    5 => "h5",
                    _ => "h6",
                }
            }
            Tag::BlockQuote => "blockquote",
            Tag::OrderedList => "ol",
            Tag::UnorderedList => "ul",
            Tag::ListItem => "li",
            Tag::FootnoteDefinition => "div",
            Tag::Table => "table",
            Tag::TableHeader => "thead",
            Tag::TableBody => "tbody",
            Tag::TableRow => "tr",
            Tag::TableHead => "th",
            Tag::TableCell => "td",
            Tag::Emphasis => "em",
            Tag::EmphasisAlt => {
                if self.options.render_underlines {
                    "u"
                } else {
                    "em"
                }
            }
            Tag::Strong => "strong",
            Tag::Strikethrough => "ss",
            Tag::Link => "a",
            Tag::Container => "div",
            Tag::Span => "span",
        }
    }

    fn start_tag(&mut self, tag: Tag, attrs: &Attrs) -> Result<(), io::Error> {
        let html_tag = self.tag_to_html_tag(tag);
        write!(self.out, "<{}", html_tag)?;

        match attrs.start {
            None | Some(1) => {}
            Some(start) => {
                write!(self.out, " start={}", start)?;
            }
        }

        let mut combined_style = String::new();
        if let Some(ref id) = attrs.id {
            write!(self.out, " id=\"{}\"", escape(id.as_str()))?;
        }
        if let Some(ref title) = attrs.title {
            write!(self.out, " title=\"{}\"", escape(title.as_str()))?;
        }
        if let Some(ref target) = attrs.target {
            write!(self.out, " href=\"{}\"", escape(target.as_str()))?;
        }

        combined_style.push_str(match attrs.alignment {
            Alignment::None => "",
            Alignment::Left => "text-align: left",
            Alignment::Center => "text-align: center",
            Alignment::Right => "text-align: right",
        });

        let mut combined_class = Cow::Borrowed("");
        if let Some(ref custom) = attrs.custom {
            for (key, value) in custom.iter() {
                if key == "style" {
                    if !combined_style.is_empty() {
                        combined_style.push_str("; ");
                    }
                    combined_style.push_str(value.as_str());
                } else if key == "class" {
                    combined_class = Cow::Borrowed(value.as_str());
                } else {
                    write!(self.out, " {}=\"{}\"", key, escape(value.as_str()))?;
                }
            }
        }

        if tag == Tag::FootnoteDefinition {
            combined_class = if combined_class.is_empty() {
                Cow::Borrowed(&self.options.footnote_definition_class)
            } else {
                Cow::Owned(format!(
                    "{} {}",
                    combined_class, &self.options.footnote_definition_class
                ))
            };
        }

        if !combined_class.is_empty() {
            write!(self.out, " class=\"{}\"", escape(&combined_class))?;
        }

        if !combined_style.is_empty() {
            write!(self.out, " style=\"{}\"", escape(&combined_style))?;
        }

        write!(self.out, ">")?;

        Ok(())
    }

    fn end_tag(&mut self, tag: Tag) -> Result<(), io::Error> {
        let html_tag = self.tag_to_html_tag(tag);

        write!(
            self.out,
            "</{}>{}",
            html_tag,
            if self.is_block_tag(tag) { "\n" } else { "" }
        )?;

        Ok(())
    }

    /// Feeds a single event into the renderer.
    pub fn feed_event(&mut self, event: &AnnotatedEvent<'data>) -> Result<(), io::Error> {
        match event.event {
            Event::DocumentStart(_) | Event::MetaData(_) => {}
            Event::StartTag(StartTagEvent { tag, ref attrs }) => {
                self.start_tag(tag, attrs)?;
            }
            Event::EndTag(EndTagEvent { tag }) => {
                self.end_tag(tag)?;
            }
            Event::Text(TextEvent { ref text }) => {
                write!(self.out, "{}", escape(text.as_str()))?;
            }
            Event::CodeBlock(CodeBlockEvent {
                ref code,
                ref language,
            }) => {
                write!(self.out, "<pre><code")?;
                if let Some(language) = language {
                    write!(self.out, " class=\"lang-{}\"", language.as_str())?;
                }
                writeln!(self.out, ">{}</code></pre>", escape(code.as_str()))?;
            }
            Event::Directive(DirectiveEvent {
                ref name, ref body, ..
            }) => {
                write!(
                    self.out,
                    "<div class=\"directive-{}\"><pre>{}</pre></div>",
                    escape(name.as_str()),
                    escape(body.as_str()),
                )?;
            }
            Event::InterpretedText(InterpretedTextEvent { ref text, ref role }) => {
                write!(
                    self.out,
                    "<span class=\"role-{}\">{}</span>",
                    escape(role.as_str()),
                    escape(text.as_str())
                )?;
            }
            Event::InlineCode(InlineCodeEvent { ref code }) => {
                write!(self.out, "<code>{}</code>", escape(code.as_str()))?;
            }
            Event::Image(ImageEvent {
                ref target,
                ref alt,
                ref title,
            }) => {
                write!(
                    self.out,
                    "<img src=\"{}\" alt=\"{}\" title=\"{}\">",
                    target,
                    alt.as_ref().map_or("", |x| x.as_str()),
                    title.as_ref().map_or("", |x| x.as_str()),
                )?;
            }
            Event::RawHtml(RawHtmlEvent { ref html }) => {
                write!(self.out, "{}", html)?;
            }
            Event::SoftBreak => writeln!(self.out)?,
            Event::HardBreak => writeln!(self.out, "<br>")?,
            Event::Rule => {
                write!(self.out, "<hr>")?;
            }
            Event::Checkbox(CheckboxEvent { checked }) => {
                write!(
                    self.out,
                    "<input type=checkbox disabled{}>",
                    if checked { " checked" } else { "" }
                )?;
            }
            Event::FootnoteReference(FootnoteReferenceEvent { ref target }) => {
                let number = match self.footnotes.get(target) {
                    Some(&num) => num,
                    None => {
                        let next_number = self.footnotes.len() + 1;
                        self.footnotes.insert(target.clone(), next_number);
                        next_number
                    }
                };
                write!(
                    self.out,
                    "<sup class=\"{}\"><a href=\"#{}\">{}</a></sup>",
                    escape(&self.options.footnote_reference_class),
                    escape(target.as_str()),
                    number,
                )?;
            }
            Event::Error(ErrorEvent {
                ref title,
                ref description,
            }) => {
                write!(
                    self.out,
                    "<div class=\"error\">\n<h3>{}</h3>\n<p>{}</p>\n</div>",
                    escape(title.as_str()),
                    escape(description.as_ref().map_or("No details", |x| x.as_str())),
                )?;
            }
        }
        Ok(())
    }

    /// Feeds an event stream into the renderer.
    pub fn feed_stream<I>(&mut self, iter: I) -> Result<(), io::Error>
    where
        I: Iterator<Item = AnnotatedEvent<'data>>,
    {
        for event in iter {
            self.feed_event(&event)?;
        }
        Ok(())
    }
}

impl<'data, 'options> HtmlRenderer<'data, 'options, Vec<u8>> {
    /// Creates a new html renderer writing into a buffer.
    pub fn new_buffered(options: &'options HtmlRendererOptions) -> Self {
        HtmlRenderer::new(Vec::new(), options)
    }

    /// Converts the renderer into a string.
    pub fn into_string(self) -> String {
        unsafe { String::from_utf8_unchecked(self.into_writer()) }
    }
}

/// Convenience shortcut that renders an event stream into HTML.
pub fn to_html<'a, I: Iterator<Item = AnnotatedEvent<'a>>>(
    iter: I,
    options: &HtmlRendererOptions,
) -> String {
    let mut renderer = HtmlRenderer::new_buffered(options);
    renderer.feed_stream(iter).unwrap();
    renderer.into_string()
}
