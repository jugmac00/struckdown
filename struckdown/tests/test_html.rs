use std::fs;

use struckdown::html::to_html;
use struckdown::parser::parse;

#[test]
fn test_basics() {
    insta::glob!("inputs/*.md", |file| {
        let source = fs::read_to_string(file).unwrap();
        let html = to_html(parse(&source));
        insta::assert_snapshot!(html);
    });
}
