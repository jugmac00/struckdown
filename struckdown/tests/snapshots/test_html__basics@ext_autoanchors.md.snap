---
source: struckdown/tests/test_html.rs
expression: html
input_file: struckdown/tests/inputs/ext_autoanchors.md
---
<h1>Heading 1</h1>
<p>Text 1</p>
<h1 id="manual-id">Heading 2</h1>
<p>Text 2</p>

